### PR TITLE
Recsys map

### PR DESCRIPTION
**Pull request by zkid18**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .` ? 
- [x] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

### CODE REVIEW

File: catalyst/metrics/avg_precision.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def avg_precision(
+    outputs: torch.Tensor, targets: torch.Tensor, k=100
+) -> torch.Tensor:
+    """
+    Calculate the Average Precision (AP) for recsys
+    The precision metric summarizes the fraction of relevant items
+    out of the whole the recommendation list.
+    The average precision @ k (AP@k) metrics summarizes the average
```
Comment (Scitator) [inline catalyst/metrics/avg_precision.py]: maybe, this could help a bit https://stackoverflow.com/questions/45534757/how-to-use-inline-code-with-a-trailing-whitespace

File: catalyst/metrics/avg_precision.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+"""
+MAP metric.
+"""
+
+import torch
+
+
+def avg_precision(
+    outputs: torch.Tensor, targets: torch.Tensor, k=100
```
Comment (Scitator) [inline catalyst/metrics/avg_precision.py]: `k=10`?

File: catalyst/metrics/tests/test_avg_precision.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import numpy as np
```
Comment (Scitator) [inline catalyst/metrics/tests/test_avg_precision.py:1]: could we also add tests from recsys course?

File: catalyst/metrics/avg_precision.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    true_sorted_by_pred_topk = true_sorted_by_preds[:, :k]
+    precisions = torch.zeros_like(true_sorted_by_pred_topk)
+
+    for index in range(k):
+        precisions[:, index] = torch.sum(
+            true_sorted_by_pred_topk[:, : (index + 1)], dim=1
+        ) / float(index + 1)
+
+    ap = torch.mean(precisions, dim=1)
+    return ap
```
Comment (Scitator) [inline catalyst/metrics/avg_precision.py:71]: could we add docs, that we return Tensor with shape `[bs]`?

Comment (Scitator) [reply catalyst/metrics/avg_precision.py:71]: and could we also add `mean_avg_precision` metric here?

**Review by Scitator — DISMISSED** [review]

**Review by Scitator — DISMISSED** [review]

File: catalyst/metrics/ndcg.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+Discounted Cumulative Gain metrics:
+    * :func:`dcg`
+    * :func:`ndcg`
+"""
+from typing import List
+
+import torch
+
+
+def dcg(
```
Comment (Scitator) [inline catalyst/metrics/ndcg.py]: not sure if we need it here :)

File: catalyst/metrics/tests/test_ndcg.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import math
+
+import numpy as np
+
+import torch
```
Comment (Scitator) [inline catalyst/metrics/tests/test_ndcg.py]: and this one :)

**Review by Scitator — DISMISSED** [review]

File: catalyst/metrics/ndcg.py (hunk 5 above)
Comment (zkid18) [reply catalyst/metrics/ndcg.py]: extra spaces or `dcg` method?
